src/daemon/rpmostreed-transaction-livefs.c

```diff
@@ -210,12 +209,34 @@ copy_new_config_files (OstreeRepo          *repo,
       etc_co_opts.sepolicy_prefix = etc_path;
 
       const char *sub_etc_relpath = etc_path + strlen ("/etc/");
+      /* We keep track of added subdirectories and skip children of it, since
+       * both the diff and checkout are recursive, but we only need to checkout
+       * the directory, which will get all children. To do better I'd say we
+       * should add an option to ostree_repo_diff() to avoid recursing into
+       * changed subdirectories.  But at the scale we're dealing with here the
+       * constants for this O(N²) algorithm are tiny.
+       */
+      for (guint j = 0; j < added_subdirs->len; j++)
+        {
+          const char *already_added_subdir = added_subdirs->pdata[j];
+          if (g_str_has_prefix (sub_etc_relpath, already_added_subdir))
+            continue;
+        }
+
       g_autoptr(GFileInfo) finfo = g_file_query_info (added_f, "standard::type",
                                                       G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
                                                       cancellable, error);
       if (!finfo)
         return FALSE;
 
+      /* If this is a directory, add it to our "added subdirs" set. See above.
+       * Yes, we're playing a bit fast-and-loose by not doing a strdup, but
+       * we know ultimately that `path` will be alive as long as the file.
+       */
+      const gboolean is_dir = g_file_info_get_file_type (finfo) == G_FILE_TYPE_DIRECTORY;
+      if (is_dir)
+        g_ptr_array_add (added_subdirs, (char*)sub_etc_relpath);
```

Comment:
As a general rule, we shouldn't [use alloca in a loop; see [here](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/issues/1307) and [here](https://hownot2code.com/2016/05/25/do-not-call-the-alloca-function-inside-loops/).  I pushed a fixup which just uses malloc; :arrow_down: